src/buildercore/cfngen.py

```diff
@@ -711,7 +711,12 @@ def regenerate_stack(stackname, **more_context):
     download_cloudformation_template(stackname)
     (pname, instance_id) = core.parse_stackname(stackname)
     more_context['stackname'] = stackname # TODO: purge this crap
-    more_context['alt-config'] = instance_id
+    # lsh@2019-09-27: usage of `instance_id` here is wrong. `instance_id` looks like "foobar" in "journal--foobar"
+    # and is only correct when an alt-config matches. We typically have alt-configs for our common environments, like
+    # ci, end2end, prod, continuumtest and has thus worked stably for a while now.
+    # ad-hoc instances whose instance-id does not match an environment will have it's alt-config ignored.
+    # the alt-config used during instance creation is found in `current_context` (but may not have always been the case).
+    more_context['alt-config'] = current_context.get('alt-config', instance_id)
```

Comment:
can give an headache, just take some sample stacks that fall into the category of custom `instance_id`s and test that `update_infrastructure` doesn't get crazy and shows an empty diff

Reply:
wow - I've really gone down the rabbit hole on this one.

Uncovered a few bugs, fixed a few things but I can't figure out why the dummy fixtures aren't including 'allow-major-version-upgrades' in the final project configuration. I have to stop it here for the day because I've spent too much time on it.